rpmio/rpmkeyring.cc

```diff
@@ -133,16 +133,32 @@ int rpmKeyringModify(rpmKeyring keyring, rpmPubkey key, rpmKeyringModifyMode mod
 	if (item->second->fp == key->fp)
 	    break;
     }
-    if (item != range.second && mode == RPMKEYRING_DELETE) {
+    if (item != range.second && (mode == RPMKEYRING_DELETE || mode == RPMKEYRING_REPLACE)) {
+	/* remove sub keys */
+	for (auto it = keyring->keys.begin(); it != keyring->keys.end();) {
+	    if (it->second->primarykey == item->second) {
+		rpmPubkeyFree(it->second);
+		it = keyring->keys.erase(it);
+	    } else {
+		++it;
```

Comment:
Don't use a for-loop when you mess with the "counter" from inside, I'm not sure that's defined behavior at all. And even if the compiler handles it, the reader wont expect it. A while-loop alarts the reader to some manual loop management ahead, see eg freeArgs() in macros.c

Reply:
Well in C `for` loops are `while` loops with fancy decorations. I find using a for loop much more clear as it indicates that we are looping over all items in order. If I see a while loop this could do anything.

Not that I care much, changed.

Reply:
Oh but that's just the thing, this loop does "anything" because it's deleting items that it's looping on, so it's far from your average "loop all items" thing.

Actually the nicer way to do this would be std::remove_if() - there's so much stuff in the STL that we haven't even scratched the surface yet :smile: